### PR TITLE
PUBDEV-6863: XGBoost POJO

### PR DESCRIPTION
**Pull request by honzasterba**
GBTree POJO example:

```/*
  Licensed under the Apache License, Version 2.0
    http://www.apache.org/licenses/LICENSE-2.0.html

  AUTOGENERATED BY H2O at 2019-09-24T12:06:23.128+02:00
  3.26.0.99999
  
  Standalone prediction code with sample test data for XGBoostModel named XGBoost_model_python_1569319545135_2

  How to download, compile and execute:
      mkdir tmpdir
      cd tmpdir
      curl http://192.168.59.167:54321/3/h2o-genmodel.jar > h2o-genmodel.jar
      curl http://192.168.59.167:54321/3/Models.java/XGBoost_model_python_1569319545135_2 > XGBoost_model_python_1569319545135_2.java
      javac -cp h2o-genmodel.jar -J-Xmx2g -J-XX:MaxPermSize=128m XGBoost_model_python_1569319545135_2.java

     (Note:  Try java argument -XX:+PrintCompilation to show runtime JIT compiler behavior.)
*/
import java.util.Map;
import hex.genmodel.GenModel;
import hex.genmodel.annotations.ModelPojo;

@ModelPojo(name="XGBoost_model_python_1569319545135_2", algorithm="xgboost")
public class XGBoost_model_python_1569319545135_2 extends GenModel {
  public hex.ModelCategory getModelCategory() { return hex.ModelCategory.Regression; }

  public boolean isSupervised() { return true; }
  public int nclasses() { return 1; }

  // Names of columns used by model.
  public static final String[] NAMES = NamesHolder_XGBoost_model_python_1569319545135_2.VALUES;

  // Column domains. The last array contains domain of response column.
  public static final String[][] DOMAINS = new String[][] {
    /* AGE */ null,
    /* RACE */ null,
    /* DPROS */ null,
    /* DCAPS */ null,
    /* CAPSULE */ null
  };
  // Prior class distribution
  public static final double[] PRIOR_CLASS_DISTRIB = null;
  // Class distribution used for model building
  public static final double[] MODEL_CLASS_DISTRIB = null;

  public XGBoost_model_python_1569319545135_2() { super(NAMES,DOMAINS,"CAPSULE"); }
  public String getUUID() { return Long.toString(-1863213802227617524L); }

  // Pass in data in a double[], pre-aligned to the Model's requirements.
  // Jam predictions into the preds[] array; preds[0] is reserved for the
  // main prediction (class for classifiers or value for regression),
  // and remaining columns hold a probability distribution for classifiers.
  public final double[] score0( double[] data, double[] preds ) {
    preds[0] =
      XGBoost_model_python_1569319545135_2_Tree_g_0_t_0.score0(data) + 
      0.5f;
    preds[0] = (float) preds[0];
    return preds;
  }
}
// The class representing training column names
class NamesHolder_XGBoost_model_python_1569319545135_2 implements java.io.Serializable {
  public static final String[] VALUES = new String[4];
  static {
    NamesHolder_XGBoost_model_python_1569319545135_2_0.fill(VALUES);
  }
  static final class NamesHolder_XGBoost_model_python_1569319545135_2_0 implements java.io.Serializable {
    static final void fill(String[] sa) {
      sa[0] = "AGE";
      sa[1] = "RACE";
      sa[2] = "DPROS";
      sa[3] = "DCAPS";
    }
  }
}

class XGBoost_model_python_1569319545135_2_Tree_g_0_t_0 {
  static float score0(double[] data) {
    return     ((Double.isNaN(data[2]) || data[2] < 2.5f) ?
      ((Double.isNaN(data[2]) || data[2] < 1.5f) ?
        ((Double.isNaN(data[0]) || data[0] < 76.5f) ?
          -0.110958904f
          :
          ((Double.isNaN(data[0]) || data[0] < 77.5f) ?
            0.0375f
            :
            -0.10000001f
          )
        )
        :
        ((Double.isNaN(data[0]) || data[0] < 58.5f) ?
          ((Double.isNaN(data[0]) || data[0] < 54.5f) ?
            ((Double.isNaN(data[1]) || data[1] < 1.5f) ?
              ((Double.isNaN(data[0]) || data[0] < 51.5f) ?
                -0.0f
                :
                ((Double.isNaN(data[0]) || data[0] < 53.5f) ?
                  -0.112500004f
                  :
                  -0.0375f
                )
              )
              :
              0.075f
            )
            :
            0.13333334f
          )
          :
          ((Double.isNaN(data[0]) || data[0] < 72.5f) ?
            ((Double.isNaN(data[0]) || data[0] < 69.5f) ?
              ((Double.isNaN(data[0]) || data[0] < 63.5f) ?
                ((Double.isNaN(data[1]) || data[1] < 1.5f) ?
                  ((Double.isNaN(data[0]) || data[0] < 59.5f) ?
                    ((Double.isNaN(data[3]) || data[3] < 1.5f) ?
                      -0.120000005f
                      :
                      0.075f
                    )
                    :
                    ((Double.isNaN(data[0]) || data[0] < 62.5f) ?
                      ((Double.isNaN(data[3]) || data[3] < 1.5f) ?
                        -0.0f
                        :
                        -0.075f
                      )
                      :
                      -0.060000002f
                    )
                  )
                  :
                  0.075f
                )
                :
                ((Double.isNaN(data[0]) || data[0] < 65.5f) ?
                  -0.11538462f
                  :
                  ((Double.isNaN(data[0]) || data[0] < 68.5f) ?
                    ((Double.isNaN(data[0]) || data[0] < 67.5f) ?
                      ((Double.isNaN(data[0]) || data[0] < 66.5f) ?
                        -0.0f
                        :
                        -0.112500004f
                      )
                      :
                      -0.0f
                    )
                    :
                    -0.120000005f
                  )
                )
              )
              :
              ((Double.isNaN(data[0]) || data[0] < 70.5f) ?
                0.056250002f
                :
                ((Double.isNaN(data[3]) || data[3] < 1.5f) ?
                  ((Double.isNaN(data[1]) || data[1] < 1.5f) ?
                    -0.023076925f
                    :
                    -0.075f
                  )
                  :
                  0.075f
                )
              )
            )
            :
            -0.097500004f
          )
        )
      )
      :
      ((Double.isNaN(data[3]) || data[3] < 1.5f) ?
        ((Double.isNaN(data[0]) || data[0] < 52.0f) ?
          -0.112500004f
          :
          ((Double.isNaN(data[0]) || data[0] < 75.5f) ?
            ((Double.isNaN(data[2]) || data[2] < 3.5f) ?
              ((Double.isNaN(data[0]) || data[0] < 73.5f) ?
                ((Double.isNaN(data[0]) || data[0] < 69.5f) ?
                  ((Double.isNaN(data[0]) || data[0] < 67.5f) ?
                    ((Double.isNaN(data[0]) || data[0] < 59.0f) ?
                      ((Double.isNaN(data[0]) || data[0] < 55.5f) ?
                        0.060000002f
                        :
                        0.01875f
                      )
                      :
                      ((Double.isNaN(data[0]) || data[0] < 64.5f) ?
                        -0.053571433f
                        :
                        0.015000001f
                      )
                    )
                    :
                    0.075f
                  )
                  :
                  ((Double.isNaN(data[0]) || data[0] < 70.5f) ?
                    -0.075f
                    :
                    ((Double.isNaN(data[0]) || data[0] < 72.5f) ?
                      ((Double.isNaN(data[0]) || data[0] < 71.5f) ?
                        -0.0f
                        :
                        0.075f
                      )
                      :
                      -0.0375f
                    )
                  )
                )
                :
                ((Double.isNaN(data[0]) || data[0] < 74.5f) ?
                  0.112500004f
                  :
                  -0.0f
                )
              )
              :
              ((Double.isNaN(data[0]) || data[0] < 59.5f) ?
                -0.112500004f
                :
                ((Double.isNaN(data[0]) || data[0] < 66.5f) ?
                  ((Double.isNaN(data[1]) || data[1] < 1.5f) ?
                    0.112500004f
                    :
                    -0.0f
                  )
                  :
                  ((Double.isNaN(data[0]) || data[0] < 70.5f) ?
                    ((Double.isNaN(data[0]) || data[0] < 68.5f) ?
                      ((Double.isNaN(data[1]) || data[1] < 1.5f) ?
                        0.0375f
                        :
                        -0.075f
                      )
                      :
                      0.075f
                    )
                    :
                    -0.0f
                  )
                )
              )
            )
            :
            ((Double.isNaN(data[0]) || data[0] < 78.0f) ?
              -0.112500004f
              :
              0.075f
            )
          )
        )
        :
        ((Double.isNaN(data[0]) || data[0] < 56.5f) ?
          -0.075f
          :
          ((Double.isNaN(data[1]) || data[1] < 1.5f) ?
            ((Double.isNaN(data[0]) || data[0] < 71.5f) ?
              0.1375f
              :
              ((Double.isNaN(data[0]) || data[0] < 72.5f) ?
                ((Double.isNaN(data[2]) || data[2] < 3.5f) ?
                  -0.075f
                  :
                  0.075f
                )
                :
                0.112500004f
              )
            )
            :
            ((Double.isNaN(data[0]) || data[0] < 59.5f) ?
              0.075f
              :
              ((Double.isNaN(data[0]) || data[0] < 64.0f) ?
                -0.075f
                :
                ((Double.isNaN(data[0]) || data[0] < 67.5f) ?
                  0.075f
                  :
                  -0.0f
                )
              )
            )
          )
        )
      )
    );
  }
}

```

GBLinear POJO Example:

```/*
  Licensed under the Apache License, Version 2.0
    http://www.apache.org/licenses/LICENSE-2.0.html

  AUTOGENERATED BY H2O at 2019-09-24T12:06:27.896+02:00
  3.26.0.99999
  
  Standalone prediction code with sample test data for XGBoostModel named XGBoost_model_python_1569319545135_6

  How to download, compile and execute:
      mkdir tmpdir
      cd tmpdir
      curl http://192.168.59.167:54321/3/h2o-genmodel.jar > h2o-genmodel.jar
      curl http://192.168.59.167:54321/3/Models.java/XGBoost_model_python_1569319545135_6 > XGBoost_model_python_1569319545135_6.java
      javac -cp h2o-genmodel.jar -J-Xmx2g -J-XX:MaxPermSize=128m XGBoost_model_python_1569319545135_6.java

     (Note:  Try java argument -XX:+PrintCompilation to show runtime JIT compiler behavior.)
*/
import java.util.Map;
import hex.genmodel.GenModel;
import hex.genmodel.annotations.ModelPojo;

@ModelPojo(name="XGBoost_model_python_1569319545135_6", algorithm="xgboost")
public class XGBoost_model_python_1569319545135_6 extends GenModel {
  public hex.ModelCategory getModelCategory() { return hex.ModelCategory.Regression; }

  public boolean isSupervised() { return true; }
  public int nclasses() { return 1; }

  // Names of columns used by model.
  public static final String[] NAMES = NamesHolder_XGBoost_model_python_1569319545135_6.VALUES;

  // Column domains. The last array contains domain of response column.
  public static final String[][] DOMAINS = new String[][] {
    /* AGE */ null,
    /* RACE */ null,
    /* DPROS */ null,
    /* DCAPS */ null,
    /* CAPSULE */ null
  };
  // Prior class distribution
  public static final double[] PRIOR_CLASS_DISTRIB = null;
  // Class distribution used for model building
  public static final double[] MODEL_CLASS_DISTRIB = null;

  public XGBoost_model_python_1569319545135_6() { super(NAMES,DOMAINS,"CAPSULE"); }
  public String getUUID() { return Long.toString(-1930148559111362004L); }

  // Pass in data in a double[], pre-aligned to the Model's requirements.
  // Jam predictions into the preds[] array; preds[0] is reserved for the
  // main prediction (class for classifiers or value for regression),
  // and remaining columns hold a probability distribution for classifiers.
  public final double[] score0( double[] data, double[] preds ) {
    preds[0] =
      (Double.isNaN(data[0]) ? 0 : (-0.0012394977f * data[0])) + 
      (Double.isNaN(data[1]) ? 0 : (-0.010776395f * data[1])) + 
      (Double.isNaN(data[2]) ? 0 : (0.048697133f * data[2])) + 
      (Double.isNaN(data[3]) ? 0 : (0.0077968906f * data[3])) + 
      -0.11220723f +
      0.5f;
    preds[0] = (float) preds[0];
    return preds;
  }
}
// The class representing training column names
class NamesHolder_XGBoost_model_python_1569319545135_6 implements java.io.Serializable {
  public static final String[] VALUES = new String[4];
  static {
    NamesHolder_XGBoost_model_python_1569319545135_6_0.fill(VALUES);
  }
  static final class NamesHolder_XGBoost_model_python_1569319545135_6_0 implements java.io.Serializable {
    static final void fill(String[] sa) {
      sa[0] = "AGE";
      sa[1] = "RACE";
      sa[2] = "DPROS";
      sa[3] = "DCAPS";
    }
  }
}

```

### CODE REVIEW

File: h2o-core/src/main/java/water/util/SBPrintStream.java [hunk 1]
```diff
@@ -208,6 +208,74 @@ public SBPrintStream toJavaStringInit(float[] ss) {
     }
     return p('}');
   }
+  
+  private final static int METHOD_SIZE_LIMIT = 10240;
```
Comment (honzasterba) [inline h2o-core/src/main/java/water/util/SBPrintStream.java]: java method size limit is 64K of bytecode, 10k array elements has proven to be a good size for split

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostModel.java [hunk 2]
```diff
@@ -591,14 +591,14 @@ private static void constructSubgraph(final RegTreeNode[] xgBoostNodes, final Sh
   }
 
 
-  private final int getXGBoostClassIndex(final String treeClass) {
+  private int getXGBoostClassIndex(final String treeClass) {
     final ModelCategory modelCategory = _output.getModelCategory();
     if(ModelCategory.Regression.equals(modelCategory) && (treeClass != null && !treeClass.isEmpty())){
       throw new IllegalArgumentException("There should be no tree class specified for regression.");
     }
-    if ((treeClass == null || treeClass.isEmpty()) && ModelCategory.Regression.equals(modelCategory)) return 0;
```
Comment (honzasterba) [inline h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostModel.java]: unrelated code simplification, I could not help myself

File: h2o-py/tests/testdir_algos/xgboost/pyunit_xgboost_pojo.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
+import sys
+import os
+sys.path.insert(1, "../../../")
+from h2o.estimators.xgboost import *
+from tests import pyunit_utils
+
+
+def compare_preds(train, test, x, y, booster, ntrees, max_depth, max_error):
```
Comment (michalkurka) [inline h2o-py/tests/testdir_algos/xgboost/pyunit_xgboost_pojo.py]: this is good, are there enough `testJavaScoring` tests in XGBoostTest to cover comparison of mojo / pojo / h2o?

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostModel.java [hunk 4]
```diff
@@ -613,4 +613,29 @@ private final int getXGBoostClassIndex(final String treeClass) {
     return treeClassIndex;
   }
 
+  //--------------------------------------------------------------------------------------------------------------------
+  // Serialization into a POJO
+  //--------------------------------------------------------------------------------------------------------------------
+
+  @Override
+  protected boolean toJavaCheckTooBig() {
+    return _output == null || _output._ntrees * _parms._max_depth > 1000;
```
Comment (Pscheidl) [inline h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostModel.java]: The line drawn here is arbitrary ?

File: h2o-py/tests/testdir_algos/xgboost/pyunit_xgboost_pojo.py (hunk 3 above)
Comment (honzasterba) [reply h2o-py/tests/testdir_algos/xgboost/pyunit_xgboost_pojo.py]: it actually discovered some bugs in the pojo, so yes, there is good coverage there as well (especially for gbtree)

**Comment by honzasterba** [comment]
all the bugs should be gone, waiting for jenkins to confirm

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostModel.java (hunk 4 above)
Comment (honzasterba) [reply h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostModel.java]: yes pretty much - something like what constitutes an amount of code a human would go through
